astropy/table/meta.py

```diff
@@ -0,0 +1,336 @@
+            in the PyYAML library to allow the required key ordering via the
+            ColumnOrderList object.  The Python 3 version insists on turning the
+            items() mapping into a list object and sorting, which results in
+            alphabetical order for the column keys.
+            """
+            value = []
+            node = yaml.MappingNode(tag, value, flow_style=flow_style)
+            if self.alias_key is not None:
+                self.represented_objects[self.alias_key] = node
+            best_style = True
+            if hasattr(mapping, 'items'):
+                mapping = mapping.items()
+                if hasattr(mapping, 'sort'):
+                    mapping.sort()
+                else:
+                    mapping = list(mapping)
+                    try:
+                        mapping = sorted(mapping)
+                    except TypeError:
+                        pass
+
+            for item_key, item_value in mapping:
+                node_key = self.represent_data(item_key)
+                node_value = self.represent_data(item_value)
+                if not (isinstance(node_key, yaml.ScalarNode) and not node_key.style):
+                    best_style = False
+                if not (isinstance(node_value, yaml.ScalarNode) and not node_value.style):
+                    best_style = False
+                value.append((node_key, node_value))
+            if flow_style is None:
+                if self.default_flow_style is not None:
+                    node.flow_style = self.default_flow_style
+                else:
+                    node.flow_style = best_style
+            return node
+
+    TableDumper.add_representer(OrderedDict, _repr_odict)
+    TableDumper.add_representer(ColumnDict, _repr_column_dict)
+
+    header = copy.copy(header)  # Don't overwrite original
+    header['datatype'] = [_get_col_attributes(col) for col in header['cols']]
+    del header['cols']
+
+    lines = yaml.dump(header, Dumper=TableDumper).splitlines()
+    return lines
+
+
+class YamlParseError(Exception):
+    pass
+
+
+def get_header_from_yaml(lines):
+    """
+    Get a header dict from input ``lines`` which should be valid YAML in the
+    ECSV meta format.  This input will typically be created by
+    get_yaml_from_header.  The output is a dictionary which describes all the
+    table and column meta.
+
+    The get_cols() method in the io/ascii/ecsv.py file should be used as a
+    guide to using the information when constructing a table using this
+    header dict information.
+
+    Parameters
+    ----------
+    lines : list
+        List of text lines with YAML header content
+
+    Returns
+    -------
+    header : dict
+        Dictionary describing table and column meta
+    """
+
+    import textwrap
+    try:
+        import yaml
+    except ImportError:
+        raise ImportError('`import yaml` failed, PyYAML package is required for ECSV format')
```

Comment:
This reminds me how much I want to revisit #476 and come up with a way to do this.  #3389 might also help with that.  Sorry, I'm getting off topic though...